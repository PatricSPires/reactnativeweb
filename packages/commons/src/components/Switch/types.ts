import { ThemeType } from '../../styles/types'

export type SwitchProps = {
  checked?: boolean
  disabled?: boolean
  onPress?: () => void
  theme?: ThemeType
  size?: string
}
