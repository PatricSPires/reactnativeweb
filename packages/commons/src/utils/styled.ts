import styleSheetSerializer from 'jest-styled-components/src/styleSheetSerializer'

export function enableStyledComponentsSerializer() {
  expect.addSnapshotSerializer(styleSheetSerializer)
}
