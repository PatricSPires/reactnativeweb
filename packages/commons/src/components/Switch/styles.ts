import styled, { css } from 'styled-components'
import { TouchableOpacity, TouchableWithoutFeedback } from 'react-native'
import { SwitchProps } from './types'
import theme from '../../styles/theme'

export const Container = styled(TouchableOpacity)`
  position: relative;
  display: inline-block;
  width: ${(props: SwitchProps) => (props.size === 'small' ? 32 : 52)}px;
  height: ${(props: SwitchProps) => (props.size === 'small' ? 16 : 24)}px;
  border-radius: ${(props: SwitchProps) =>
    props.size === 'small' ? 100 : 12}px;
  background-color: ${theme.colors.medium_contrast_border};
  outline: 2px solid;
  outline-color: transparent;

  ${(props: SwitchProps) =>
    props.checked &&
    css`
      background-color: ${theme.supportColors.success};

      &:focus {
        outline-color: ${theme.supportColors.information};
      }
    `}

  ${(props: SwitchProps) =>
    props.disabled &&
    css`
      background-color: ${theme.colors.primary_subtle_border};
      outline-color: transparent;
    `}
`

export const Wrapper = styled(TouchableWithoutFeedback)`
  position: absolute;
  width: ${(props: SwitchProps) => (props.size === 'small' ? 10 : 18)}px;
  height: ${(props: SwitchProps) => (props.size === 'small' ? 10 : 18)}px;
  border-radius: 50%;
  background-color: #fff;
  top: 3px;
  left: 3px;
  -webkit-transition: left 0.25s ease;
  transition: left 0.25s ease;

  ${({ checked, size }: SwitchProps) =>
    checked &&
    `
    -webkit-transition: left 0.25s ease;
    transition: left 0.25s ease;
    left: ${size === 'small' ? 19 : 31}px;
  `}

  ${({ disabled }) =>
    disabled &&
    `
    background-color: ${theme.colors.medium_contrast_border};
  `}
`
