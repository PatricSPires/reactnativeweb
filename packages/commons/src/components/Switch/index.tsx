import React from 'react'
import { Text } from 'react-native'
import { SwitchProps } from './types'
import { Container, Wrapper } from './styles'

const Switch = ({ checked, size, disabled, onPress }: SwitchProps) => (
  <Container
    activeOpacity={0}
    size={size}
    checked={checked}
    disabled={disabled}
    onPress={onPress}
  >
    <Wrapper
      checked={checked}
      size={size}
      disabled={disabled}
      onPress={onPress}
    >
      <Text />
    </Wrapper>
  </Container>
)

export default Switch
