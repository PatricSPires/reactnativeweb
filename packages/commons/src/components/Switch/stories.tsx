import React from 'react'
import { Story } from '@storybook/react/types-6-0'
import { SwitchProps } from './types'
import Switch from './index'

export default {
  title: 'Formulário/Switch',
  component: Switch,
  parameters: {
    docs: {
      description: {
        component:
          'Este é o nosso componente de switch, que possui alguns estilos com base nas propriedades e estados.',
      },
    },
  },
  args: {
    checked: false,
    disabled: false,
    size: 'large',
  },
  argTypes: {
    onPress: {
      action: 'clicked',
      description: 'Função que será executada quando o componente for clicado.',
    },
    checked: {
      description: 'Alterna o estado do componente entre verdadeiro e falso.',
    },
    disabled: {
      description:
        'Altera o estado do componente para desabilitado e não permite interação do usuário.',
    },
    size: {
      description: 'Altera o tamanho do componente',
    },
  },
}

export const Basic: Story<SwitchProps> = (args) => <Switch {...args} />

export const Checked = Basic.bind({})
Checked.parameters = {
  docs: {
    description: {
      story:
        'Estado do componente quando está com a propriedade __checked__:true',
    },
  },
}

Checked.args = {
  checked: true,
}

export const CheckedAndDisabled = Basic.bind({})
CheckedAndDisabled.parameters = {
  docs: {
    description: {
      story:
        'Estado do componente quando está com a propriedade __checked__:true e __disabled__:true',
    },
  },
}

CheckedAndDisabled.args = {
  disabled: true,
  checked: true,
}
