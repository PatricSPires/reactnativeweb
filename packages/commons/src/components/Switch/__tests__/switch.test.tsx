import * as React from 'react'
import { render } from '@testing-library/react-native'
import { enableStyledComponentsSerializer } from '../../../utils/styled'

import Switch from '../index'

beforeAll(() => {
  enableStyledComponentsSerializer()
})

describe('Switch render', () => {
  it('should render the switch', () => {
    const { toJSON } = render(<Switch />)

    expect(toJSON).toMatchSnapshot()
  })
})
