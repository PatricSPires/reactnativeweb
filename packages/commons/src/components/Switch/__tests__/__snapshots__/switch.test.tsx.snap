// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Switch render should render the switch 1`] = `[Function]`;
